[feature] add static analysis tool

Treggats opened the pull request:
Introduce PHPStan as static analyser. Continues the work done on #2471

resolves #1930

This PR does the following
1. add a basic config for PHPStan
2. resolve issues found for level 1
3. add a baseline for issues found in level 2
4. add a job to the coding standard Github Workflow

Treggats committed:
add PHPStan package (#1930)

Treggats committed:
#1930 add PHPStan config

Treggats committed:
#1930 add PHPStan to the coding-standards Github Workflow

Treggats committed:
#1930 remove redundant config settings

Treggats committed:
#1930 continue job on error

Treggats committed:
#1930 add required return statements

Treggats committed:
#1930 rename arguments, keep in line with the parent name

Treggats committed:
#1930 update return type

Treggats committed:
#1930 update parameter type

Treggats committed:
#1930 remove default config option values

Treggats committed:
#1930 ignore two errors

Treggats commented on .editorconfig:
```diff
@@ -6,4 +6,7 @@ end_of_line = lf
+trim_trailing_whitespace = true
+
+[*.yml]
+indent_size = 2
```
EOL final newline
```suggestion
indent_size = 2

```

Treggats replied:
Match the current indenting.

Treggats commented on .github/workflows/build-ci.yml:
```diff
@@ -2,8 +2,6 @@ name: CI
 
 on:
     push:
-        branches:
```
redundant, and not according to the Github Workflow schema

Treggats commented on .github/workflows/coding-standards.yml:
```diff
@@ -53,3 +51,29 @@ jobs:
+
+  analysis:
+    runs-on: "ubuntu-22.04"
+    continue-on-error: true
```
This will allow the job to continue even if PHPStan finds errors.

For now mostly a convenience to see the output of runs for both PHP 8.1 and 8.2

Treggats commented on .github/workflows/coding-standards.yml:
```diff
@@ -53,3 +51,29 @@ jobs:
       - name: "Run PHP_CodeSniffer"
         run: "vendor/bin/phpcs -q --no-colors --report=checkstyle | cs2pr"
+
+  analysis:
```
Borrowed from the changes that @divine made in #2471

Treggats commented on .github/workflows/coding-standards.yml:
```diff
@@ -53,3 +51,29 @@ jobs:
+          extensions: curl, mbstring
+          tools: composer:v2
+          coverage: none
+          # ini-values: memory_limit=-1
```
No need to increase the memory limit, runs just fine without.
```suggestion

```

Treggats commented on composer.json:
```diff
@@ -31,10 +31,12 @@
         "mongodb/mongodb": "^1.15"
     },
     "require-dev": {
+        "roave/security-advisories": "dev-latest",
```
Nice to have, will cause `composer install` to halt if there is a security advisory for one of the dependencies

GromNaN replied:
This project is not directly deployed. We will not restrict versions based on known security issues. Please remove this job.

Treggats replied:
👌 
```suggestion
```

Treggats commented on phpstan-baseline.neon:
```diff
@@ -0,0 +1,46 @@
+parameters:
```
The baseline provides PHPStan with a list of errors to ignore, kind of saying 
> I know these issues exist, will fix it later

Treggats commented on phpstan.neon.dist:
```diff
@@ -0,0 +1,17 @@
+    editorUrl: 'phpstorm://open?file=%%file%%&line=%%line%%'
+
+    ignoreErrors:
+        - '#Unsafe usage of new static#'
```
These issues were the remaining errors from level 1, I wasn't sure how to fix them. So I chose to ignore them.

Treggats commented on phpunit.xml.dist:
```diff
@@ -1,13 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/10.4/phpunit.xsd"
-         backupGlobals="false"
```
removed options that had the default value.

Treggats commented on src/Relations/BelongsToMany.php:
```diff
@@ -82,11 +82,11 @@ protected function setWhere()
 
     /** @inheritdoc */
-    public function save(Model $model, array $joining = [], $touch = true)
+    public function save(Model $model, array $pivotAttributes = [], $touch = true)
```
The parent method argument name was changed, updated it to match it.

Treggats commented on .github/workflows/build-ci.yml:
```diff
@@ -2,8 +2,6 @@ name: CI
 
 on:
     push:
```
@divine should this always run on push? I think it makes more sense to only run for a pull request event.

In our application we have the following defined in our code quality Github Workflow. We run on certain pull request event and push to the develop branch. Reduces the Github Action minutes, and the double runs of the workflow.

```yml
on:

  pull_request:
    types:
      - opened
      - synchronize
      - ready_for_review

  workflow_dispatch:

  push:
    branches:
      - develop
```

divine replied:
As this is an open-source project there is no limit on minutes, so it should be fine.
The reason from what I remember is that sometimes I've pushed commits directly to the branches which of course is unnecessary.
Your suggested change looks great.

Thanks!

Treggats commented on phpstan.neon.dist:
```diff
@@ -0,0 +1,17 @@
+    - ./phpstan-baseline.neon
+
+parameters:
+    tmpDir: .cache/phpstan
```
Will speedup subsequent runs, in a Github Workflow we'd need to use the cache action to store this cache

Treggats committed:
#1930 add formatting to the coding-standards Github Workflow

Treggats commented on .github/workflows/coding-standards.yml:
```diff
@@ -53,3 +56,61 @@ jobs:
       - name: "Run PHP_CodeSniffer"
         run: "vendor/bin/phpcs -q --no-colors --report=checkstyle | cs2pr"
+
+      - name: "Format the code"
```
Nice to have, removes the need to remember to run this locally

GromNaN replied:
I can see the benefit, but I'm not sure...

Treggats replied:
Just an idea, I'm fine with removing it if need be

divine replied:
I think formatting the code should happen after each commit but this might make things ugly.

Treggats replied:
@divine keep it or remove it?

GromNaN commented on src/Relations/EmbedsMany.php:
```diff
@@ -195,9 +196,11 @@ public function destroy($ids = [])
     /**
      * Delete all embedded models.
      *
+     * @param null $id
```
Can we remove this argument, since it's not used.

Treggats replied:
I know that it's not used, but it is part of the methods heading (name, arguments, etc.) as defined in the [parent method](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Query/Builder.php#L3147).

Should we not keep the inherited methods heading the same, even though we might not use some arguments?

GromNaN replied:
Ok, then we need to add a comment to mention that the id is not used. And throw a LogicException if it is not null. We don't want to give the illusion that embeds can be deleted by Id. Even though it's an interesting feature to implement.

GromNaN replied:
The note must be on the argument description.
```suggestion
     * @param null $id Not used. Inherited from parent method.
```

GromNaN commented on src/Relations/EmbedsOne.php:
```diff
@@ -133,9 +133,11 @@ public function dissociate()
      * @return int
      */
-    public function delete()
+    public function delete($id = null)
```
Same here, can we remove this argument, since it's not used.

GromNaN commented on src/Relations/EmbedsOneOrMany.php:
```diff
@@ -42,6 +43,8 @@ abstract class EmbedsOneOrMany extends Relation
      */
     public function __construct(Builder $query, Model $parent, Model $related, string $localKey, string $foreignKey, string $relation)
     {
+        parent::__construct($query, $parent);
```
I think parent constructor should be called last. The affectation of `$this->query` and `$this->parent` can be removed, as the duplicate `$this->addConstraints();` at the end.

Treggats replied:
Do you mean that those can be removed?

GromNaN replied:
Yes, it we call the parent constructor, we don't need to repeat all it does.

GromNaN commented on src/Relations/EmbedsOneOrMany.php:
```diff
@@ -101,9 +104,11 @@ public function get($columns = ['*'])
      * @return int
      */
-    public function count()
+    public function count($columns = '*')
```
Not used.

GromNaN commented on composer.json:
```diff
@@ -31,10 +31,12 @@
         "mockery/mockery": "^1.4.4",
-        "doctrine/coding-standard": "12.0.x-dev"
+        "doctrine/coding-standard": "12.0.x-dev",
+        "nunomaduro/larastan": "^2.6"
```
I understand that larastan's interest is in its ability to boot the container to analyze the facades. We don't have any applications to boot in this library. I'd rather use phpstan directly instead of depending on an extra abstraction layer and its potential breaking changes.

Treggats replied:
Alright, I'll replace it with "native" PHPStan

Treggats committed:
#1930 rename argument names, in line with parent method

Also remove second argument, only one argument needed

Treggats committed:
#1930 updated the baseline

Treggats committed:
#1930 add baseline for level 2

Treggats committed:
#1930 add PHPStan to the coding-standards Github Workflow

Treggats committed:
#1930 assert that a instance of the Mongo query builder is used

The base `update(..)` method and the MongoDB `update(..)` differ in the amount of arguments.
Adding a assertion to make sure the correct one is used.

Treggats committed:
#1930 convert to short closure

Treggats commented on src/Eloquent/Builder.php:
```diff
@@ -71,7 +73,10 @@ public function update(array $values, array $options = [])
+        $builder = $this->toBase();
+        assert($builder instanceof MongoDBQueryBuilder);
+
+        return $builder->update($this->addUpdatedAtColumn($values), $options);
```
So this is an interesting case. Both the IDE and PHPStan think that `Illuminate\Database\Query\Builder` is used, while actually it's `MongoDB\Laravel\Query\Builder`.
Adding the assertion is quickest/eastiest way to make sure the correct one is used.

GromNaN replied:
You can overload the return type of `toBase` with an annotation on this class.
```php
/** @method \MongoDB\Laravel\Query\Builder toBase() */
class Builder extends EloquentBuilder
```

Treggats commented on src/Eloquent/Builder.php:
```diff
@@ -228,16 +233,15 @@ protected function ensureOrderForCursorPagination($shouldReverse = false)
-            $this->query->orders = collect($this->query->orders)->map(function ($direction) {
-                return $direction === 1 ? -1 : 1;
-            })->toArray();
+            $this->query->orders = collect($this->query->orders)
```
Thought that this was a bit cleaner.

Treggats committed:
#1930 add missing parent calls

Treggats committed:
#1930 add parent call and simplify setting the properties

Treggats committed:
#1930 update phpunit config using `phpunit --generate-configuration`

Treggats commented on src/Relations/EmbedsOneOrMany.php:
```diff
@@ -42,8 +43,8 @@ abstract class EmbedsOneOrMany extends Relation
-        $this->parent     = $parent;
+        parent::__construct($query, $parent);
+
         $this->related    = $related;
```
This property is also being set in the parent constructor, but when I tried to change it a few mocks broke and I'm not sure how to update those. So I left it alone 😅

Treggats commented:
Putting in draft while I figure out why the tests are faling

GromNaN commented on .github/workflows/coding-standards.yml:
```diff
@@ -53,3 +56,61 @@ jobs:
         run: "vendor/bin/phpcs -q --no-colors --report=checkstyle | cs2pr"
+
+      - name: "Format the code"
+        run: ./vendor/bin/phpcbf
```
Some errors can't be fixed. We need to run `./vendor/bin/phpcs` after this command to avoid committing partially fixed code.

Treggats committed:
#1930 increase the php memory limit

Treggats committed:
#1930 formatting/readability GH Action

Treggats committed:
#1930 remove default value

Treggats commented on .cache/.gitignore:
```diff
@@ -0,0 +1,2 @@
+*
```
make sure the directory for the cache is present

GromNaN replied:
Why can't this directory be created automatically? I would prefer not to have an empty folder in the project

Treggats replied:
I'll see if I can make that happen

Treggats commented on .github/workflows/build-ci.yml:
```diff
@@ -39,41 +38,47 @@ jobs:
             -   uses: codecov/codecov-action@v3
                 with:
                     token: ${{ secrets.CODECOV_TOKEN }}
-                    fail_ci_if_error: false
```
This is the default value, so it can be omitted

Treggats commented on .gitignore:
```diff
@@ -3,9 +3,9 @@
 *.sublime-workspace
 .DS_Store
 .idea/
-.phpunit.cache/
```
I introduced the `.cache` directory for caching phpunit/phpcs/phpstan

Treggats commented on phpunit.xml.dist:
```diff
@@ -20,10 +18,15 @@
         <env name="SQLITE_DATABASE" value=":memory:"/>
         <env name="QUEUE_CONNECTION" value="database"/>
+
+        <ini name="memory_limit" value="-1"/>
```
While running the tests locally in Docker I ran into the issue that there wasn't enough memory. Disabling the limit here, seems like the most logical place

Treggats commented on phpunit.xml.dist:
```diff
@@ -20,10 +18,15 @@
+        restrictNotices="true"
+        restrictWarnings="true">
         <include>
-            <directory suffix=".php">./src</directory>
```
the suffix was redundant so it could be removed

GromNaN commented on .github/workflows/build-ci.yml:
```diff
@@ -30,6 +28,7 @@ jobs:
 
         steps:
             -   uses: actions/checkout@v4
+
```
Could you revert this empty newlines, it forces you to scroll more than necessary. The `-` already make each task legible.

GromNaN commented on src/Eloquent/Builder.php:
```diff
@@ -228,16 +233,15 @@ protected function ensureOrderForCursorPagination($shouldReverse = false)
-                return $direction === 1 ? -1 : 1;
-            })->toArray();
+            $this->query->orders = collect($this->query->orders)
+                ->map(fn (int $direction) => $direction === 1 ? -1 : 1)
```
Good use of an arrow function.
```suggestion
                ->map(static fn (int $direction) => $direction === 1 ? -1 : 1)
```

GromNaN commented on src/Eloquent/Builder.php:
```diff
@@ -228,16 +233,15 @@ protected function ensureOrderForCursorPagination($shouldReverse = false)
-        return collect($this->query->orders)->map(function ($direction, $column) {
-            return [
+        return collect($this->query->orders)
+            ->map(fn ($direction, $column) => [
```
```suggestion
            ->map(static fn ($direction, $column) => [
```

Treggats committed:
Apply suggestions from code review

Co-authored-by: Jérôme Tamarelle <jerome@tamarelle.net>

Treggats committed:
continue workflow if `phpcbf` finds/fixes issues

Treggats committed:
apply phpcbf formatting

Treggats committed:
cs

Treggats committed:
use static call for static method

Treggats committed:
let early return return the expected data

Treggats commented on src/Relations/EmbedsOneOrMany.php:
```diff
@@ -265,17 +273,17 @@ protected function toCollection(array $records = [])
-
         $connection = $this->related->getConnection();
 
+        if (empty($attributes)) {
```
Return a actual model instance

Treggats commented on src/Relations/EmbedsMany.php:
```diff
@@ -82,7 +86,7 @@ public function performUpdate(Model $model)
         $foreignKey = $this->getForeignKeyValue($model);
 
-        $values = $this->getUpdateValues($model->getDirty(), $this->localKey . '.$.');
+        $values = self::getUpdateValues($model->getDirty(), $this->localKey . '.$.');
```
The method is static, so using a static call here

GromNaN commented on src/Relations/EmbedsMany.php:
```diff
@@ -195,10 +199,16 @@ public function destroy($ids = [])
-     * @return int
+     * @param null $id
+     *
+     * @throws LogicException|Throwable
```
I never use `@throw` annotation for logic exception, as they are made for the developer and should not be caught.

Treggats replied:
Should I remove them?

GromNaN replied:
Yes please.

Treggats commented on src/Relations/EmbedsMany.php:
```diff
@@ -195,10 +199,16 @@ public function destroy($ids = [])
+     * @param null $id
+     *
+     * @throws LogicException|Throwable
+     *
+     * @note The $id is not used to delete embedded models.
```
```suggestion
```

Treggats commented on src/Relations/EmbedsOne.php:
```diff
@@ -133,10 +137,16 @@ public function dissociate()
      * Delete all embedded models.
      *
-     * @return int
+     * @param ?string $id
+     *
+     * @throws LogicException|Throwable
+     *
+     * @note The $id is not used to delete embedded models.
```
```suggestion
     * @param ?string $id Not used. Inherited from parent method.
     *
     * @throws LogicException|Throwable
```

Treggats commented on src/Relations/EmbedsOneOrMany.php:
```diff
@@ -101,10 +103,16 @@ public function get($columns = ['*'])
      * Get the number of embedded models.
      *
-     * @return int
+     * @param Expression|string $columns
+     *
+     * @throws LogicException|Throwable
+     *
+     * @note The $column parameter is not used to count embedded models.
```
```suggestion
     * @param Expression|string $columns Not used. Inherited from parent method.
     *
     * @throws LogicException|Throwable
```

alcaeus commented on .github/workflows/build-ci.yml:
```diff
@@ -66,14 +64,11 @@ jobs:
+                run: ./vendor/bin/phpunit --coverage-clover coverage.xml
                 env:
                     MONGODB_URI: 'mongodb://127.0.0.1/?replicaSet=rs'
             -   uses: codecov/codecov-action@v3
                 with:
                     token: ${{ secrets.CODECOV_TOKEN }}
```
@GromNaN I noticed we still have this in. Given that we're no longer using codecov.io, should we remove this step?

GromNaN replied:
Yes.

alcaeus commented on .github/workflows/coding-standards.yml:
```diff
@@ -50,6 +53,67 @@ jobs:
         with:
           composer-options: "--no-suggest"
 
+      - name: "Format the code"
+        continue-on-error: true
+        run: |
+          mkdir .cache
+          ./vendor/bin/phpcbf
+
```
@GromNaN what are your thoughts on this? I'm usually not a fan of workflows that commit changes, but since it's "just" checkstyle it may be fine.

GromNaN replied:
I think it's an acceptable way of speeding up workflow when you forget to launch phpcs. It adds a commit, so it's easy to undo and will be squashed during the merge.

alcaeus reviewed (commented):
I have a few comments on the CI changes, haven't yet reviewed the changes to satisfy PHPStan.

@Treggats can you please change the target branch of this PR to 4.1 and rebase changes to resolve conflicts? I don't feel comfortable introducing these changes in a patch release.

Treggats committed:
unused, removing

The codecov.io service is no longer used.

Treggats committed:
update PHPStan baseline (level 2)

Treggats commented:
@alcaeus I've updated the base and removed the codecov.io step in the Github Action workflow

Treggats committed:
replace assert with docblock

Treggats committed:
remove `@throw` docblock comment

GromNaN commented on phpstan-baseline.neon:
```diff
@@ -0,0 +1,31 @@
+parameters:
+	ignoreErrors:
+		-
+			message: "#^Variable \\$value on left side of \\?\\? always exists and is always null\\.$#"
```
If we remove the null coalescence operator in `Model::fromJson`, tests still pass. 
https://github.com/mongodb/laravel-mongodb/blob/82ddb839ed2f3c027d4b810a66d056ba5221d4bb/src/Eloquent/Model.php#L297-L304
This was introduced in #2653, could you check this error @hans-thomas?

hans-thomas replied:
@Treggats We can remove `Null Coalescing` Operator. If the `$value` was `Null`, the ` Json::encode` will handle it.

UPDATE: To solve the second error, we can also remove the second  `Null Coalescing` Operator.

The final method:
```php
public function fromJson($value, $asObject = false) 
 { 
     if (! is_string($value)) { 
         $value = Json::encode($value); 
     } 
  
     return Json::decode($value, ! $asObject); 
 } 
```

GromNaN commented on phpstan-baseline.neon:
```diff
@@ -0,0 +1,31 @@
+		-
+			message: "#^Call to an undefined method Illuminate\\\\Database\\\\Eloquent\\\\Builder\\|Illuminate\\\\Database\\\\Eloquent\\\\Model\\:\\:pull\\(\\)\\.$#"
+			count: 1
+			path: src/Relations/BelongsToMany.php
```
This should be fixed by adding `assert($this->parent instanceof MongoDB\Eloquent\Model)`

GromNaN commented on phpstan-baseline.neon:
```diff
@@ -0,0 +1,31 @@
+			path: src/Relations/BelongsToMany.php
+
+		-
+			message: "#^Cannot call method modelKeys\\(\\) on array\\|Illuminate\\\\Support\\\\Collection\\.$#"
```
It looks like there is an hidden bug if `$ids` is an array.
https://github.com/mongodb/laravel-mongodb/blob/82ddb839ed2f3c027d4b810a66d056ba5221d4bb/src/Relations/BelongsToMany.php#L129-L132

hans-thomas replied:
https://github.com/mongodb/laravel-mongodb/blob/82ddb839ed2f3c027d4b810a66d056ba5221d4bb/src/Relations/BelongsToMany.php#L127-L132
It's not even covered in the tests. 
In what case, the `$this->parent->{$this->relatedPivotKey}` could return a `Collection`?
I think in the past, it was possible but not now. The pivot column always is an array.

UPDATE: removing this condition doesn't cause any error.

GromNaN reviewed (commented):
The *virtual* arguments of `Model::push()` must be documented to remove the errors with `Relations/MorphToMany`.

`MorphToMany::updateExistingPivot()` can return `0`, this returned value is documented but not used. https://github.com/laravel/framework/blob/2967d89906708cc7d619fc130e835c8002b7d3e3/src/Illuminate/Database/Eloquent/Relations/Concerns/InteractsWithPivotTable.php#L187

Treggats committed:
fix PHPStan error; unknown method `pull()`

Treggats committed:
replace array_merge with array unpacking

Treggats committed:
add return type to match the userland use case

Treggats committed:
ignore Eloquent magically passing arguments

Treggats committed:
replace with the suggested "new" method

Treggats commented on phpstan-baseline.neon:
```diff
@@ -0,0 +1,31 @@
+			path: src/Relations/BelongsToMany.php
+
+		-
+			message: "#^Method Illuminate\\\\Database\\\\Eloquent\\\\Model\\:\\:push\\(\\) invoked with 3 parameters, 0 required\\.$#"
```
The `push()` method uses `func_get_args()` to get its arguments, but we're passing it to the method as arguments. Not sure how to fix or write differently.

GromNaN replied:
My idea was to add phpdoc for this arguments, but PHPStorm and phpcs both complain that "PHPDoc tag `@param` references unknown parameter". 

```
    /*
     * If no arguments are provided, save the model and all of its relationships.
     * If a $column is provided, append the $values to the $column array attribute.
     *
     * @param string $column
     * @param mixed  $values
     * @param bool   $unique
     *
     * @return bool
     */
    public function push(/* string $column = null, $values = null, bool $unique = false */)
```

I'm delaying it for a later PR. No doubt the design of this method is incorrect and needs to be changed.

hans-thomas commented on src/Relations/BelongsToMany.php:
```diff
@@ -229,6 +230,8 @@ public function detach($ids = [], $touch = true)
 
         // Remove the relation to the parent.
+        assert($this->parent instanceof \MongoDB\Laravel\Eloquent\Model);
+        assert($query instanceof \MongoDB\Laravel\Eloquent\Model);
```
```suggestion
        assert($query instanceof \MongoDB\Laravel\Eloquent\Builder);
```
The `$query` variable is an instance of `Builder` class.

Treggats committed:
Apply suggestions from code review

Co-authored-by: Mohammad Mortazavi <39920372+hans-thomas@users.noreply.github.com>

hans-thomas commented on phpstan-baseline.neon:
```diff
@@ -0,0 +1,31 @@
+parameters:
+	ignoreErrors:
+		-
+			message: "#^Variable \\$value on left side of \\?\\? always exists and is always null\\.$#"
+			count: 1
+			path: src/Eloquent/Model.php
+
+		-
+			message: "#^Variable \\$value on left side of \\?\\? always exists and is not nullable\\.$#"
+			count: 1
+			path: src/Eloquent/Model.php
+
```
```suggestion

```
You can remove these two errors after applying this [change](https://github.com/mongodb/laravel-mongodb/pull/2664/files#r1403260408).

Treggats committed:
fix null coalescing

Co-authored-by: Mohammad Mortazavi <39920372+hans-thomas@users.noreply.github.com>

Treggats committed:
remove Collection check

Co-authored-by: Mohammad Mortazavi <39920372+hans-thomas@users.noreply.github.com>

Treggats commented on src/Relations/BelongsToMany.php:
```diff
@@ -295,28 +293,6 @@ public function getQualifiedRelatedPivotKeyName()
-     *
-     * @return array
-     */
-    protected function formatSyncList(array $records)
```
removing in favor of `formatRecordList` as mentioned in the comment

hans-thomas commented on src/Relations/BelongsToMany.php:
```diff
@@ -126,12 +126,7 @@ public function sync($ids, $detaching = true)
         // if they exist in the array of current ones, and if not we will insert.
         $current = $this->parent->{$this->relatedPivotKey} ?: [];
 
-        // See issue #256.
-        if ($current instanceof Collection) {
-            $current = $ids->modelKeys();
-        }
```
Could you please remove these lines in the[ `MorphToMany`](https://github.com/hans-thomas/laravel-mongodb/blob/d47380ccfb3cba967855191600847133726ceed3/src/Relations/MorphToMany.php#L136-L139) class too?

Treggats committed:
remove Collection check

Co-authored-by: Mohammad Mortazavi <39920372+hans-thomas@users.noreply.github.com>

GromNaN commented:
Thank you @Treggats

Treggats commented:
> Thank you @Treggats

You're welcome @GromNaN was a nice experience
